SourceCode/raytracer/CRTTypes.h

```diff
@@ -245,5 +245,75 @@ class Triangle {
         return "Triangle: {" + v0.toString() + ", " + v1.toString() + ", " + v2.toString() + "}";
     }
 
+    /*
+    * Input: ray, rayProj
+    * Input: rayProj: projection of ray.direction onto camera direction
+    * Output: t: distance from ray.origin to the intersection point
+    * Output: p: intersection point
+    * return: true if the ray intersects the triangle
+    */
+    bool intersect(const Ray& ray, float& t, Vec3& p) const {
+        // Here assuming counter-clockwise order
+        Vec3 e0 = v1 - v0;
+        Vec3 e1 = v2 - v0;
+        Vec3 tri_normal = cross(e0, e1).normalize();
+
+        float rayProj = dot(ray.direction, tri_normal);
+
+        // if triangle is facing the ray, compute ray-plane intersection
+        if (rayProj < -1e-6) {
+            float rpDist = dot(tri_normal, ray.origin - v0);
+            t = rpDist / rayProj;
+            p = ray.origin + ray.direction * t;
+
+            // check if `p` is inside triangle
+            Vec3 c0 = cross(e0, p - v0);
+            Vec3 c1 = cross(v2 - v1, p - v1);
+            Vec3 c2 = cross(v0 - v2, p - v2);
+
+            return dot(tri_normal, c0) > 1e-6 &&
+                   dot(tri_normal, c1) > 1e-6 &&
+                   dot(tri_normal, c2) > 1e-6;
+        }
+        return false;
+    }
+
+    bool intersect_plane(const Ray& ray, float& t, Vec3& p) const {
+        // Here assuming counter-clockwise order
+        Vec3 e0 = v1 - v0;
+        Vec3 e1 = v2 - v0;
+        Vec3 tri_normal = cross(e0, e1).normalize();
 
+        float rayProj = dot(ray.direction, tri_normal);
+
+        // if triangle is facing the ray, compute ray-plane intersection
+        if (rayProj < -1e-6) {
+            float rpDist = dot(tri_normal, v0 - ray.origin); // was this the problem?
+            t = rpDist / rayProj;
+            p = ray.origin + ray.direction * t;
+            return true;
+        }
+        return false;
+    }
+
+    bool intersect_plane2(const Ray& ray, float& t, Vec3& p) const
```

Comment:
intersect_plane2 is second try: Code copied and adapted from https://www.scratchapixel.com/lessons/3d-basic-rendering/minimal-ray-tracer-rendering-simple-shapes/ray-plane-and-ray-disk-intersection.html
Has the exact same problem